utility.hpp.in

```diff
@@ -13,17 +13,28 @@
  */
 namespace utility{
 
-/* forward Declaration of function signatures */
+/* functions */
+/** @returns A copy of the argument value */
+template< typename T >
+T   copy( const T& t ){ return t; }
```

Comment:
> This is essentially the copy equivalent of the std::move function. I assume we'll be using this when we do not want to move an argument into an rvalue reference?

Yes, specifically an lvalue to a function requiring an rvalue reference argument.

> if we use this too often, would it not be better to not use rvalue references (using a local variable instead) in a function call and explicitly use std::move when we want to move instead of copy?

I'm not sure I understand. Does the following example capture your suggestion?

Given a function with the prototype
```cpp
void foo( expensiveType&& e);
```
you are suggesting
```cpp
expansiveType bar;
auto copy = bar;
foo( std::move(copy) );
/* additional operations using bar */
```
as opposed to 
```cpp
expansiveType bar;
foo( utility::copy(bar) );
/* additional operations using bar */
```

\<myThoughts\>
The former is more expensive (without assuming certain compiler optimizations not guaranteed by the standard) and is more verbose. More over, the latter introduces a new variable to keep track of (and to name), as well as the potential to mistakenly use the `copy` variable in while still in an invalid state following the move operation. For most standard library container types, this results segmentation fault at runtime, which introduces extra pressure to test thoroughly
\</myThoughts\>

>  I think this is the more standard way (from what I see on the internet, I would assume this is the case)

In my experience, common wisdom is to specify these sort of prototypes using value semantics, e.g.
```cpp
void foo( expensiveType e);
```
and when a move is desired, call to `std::move`, otherwise the value is implicit copied. My issue with this convention is that the default behavior (with respect to lvalues) is both 

- implicit
- and expensive. 

\<myThoughts\>
In my experience, when refactoring for performance working on NJOY and elsewhere, the largest gains are generally made by eliminating unnecessary implicit data copying of lvalues because of value semantics. By retaining the function prototype accepting an rvalue reference, a developer retain the ability to copy expensive types, but can't do so accidentally.
\</myThoughts\>